common/src/main/java/org/opensearch/ml/common/connector/HttpConnector.java

```diff
@@ -264,6 +265,64 @@ public  <T> T createPredictPayload(Map<String, String> parameters) {
         return (T) parameters.get("http_body");
     }
 
+    protected String applyCustomSubstitution(Map<String, String> parameters, String inputPayload) {
+
+        if (parameters == null) {
+            throw new IllegalArgumentException("Some parameter placeholder not filled in payload: input");
+        }
+        String result = inputPayload;
+
+        Pattern pattern = Pattern.compile("\\$\\{parameters\\.(\\w+)\\}");
+        Matcher matcher = pattern.matcher(inputPayload);
+
+        while (matcher.find()) {
+            String paramName = matcher.group(1);
+            String paramValue = parameters.get(paramName);
+            if (paramValue != null) {
+                String modifiedValue = isNotPlainString(paramValue) ? paramValue : "\"" + paramValue + "\"";
+                result = result.replace("${parameters." + paramName + "}", modifiedValue);
+            }
+        }
+
+        result = StringEscapeUtils.unescapeJava(result); // Unescape Unicode escape sequences
+
+        return result;
+    }
+
+    protected boolean isNotPlainString(String value) {
```

Comment:
If user input `{"key1": "123"}`, we will detect `"123"` as not plain String?

Reply:
From line 282 
`isNotPlainString(paramValue) ? paramValue : "\"" + paramValue + "\"";`

It will return `123`  rather than `"123"` which not what user expected.

Reply:
1. So if user provides a number with `""` (`"123"`), although it's a number we still should send value as `"123"`?
2. if user provide a number without `""` (123), in that case I suppose we want to send the value as `123`?
3. What about boolean values? Same idea?

Reply:
For example "{ \"texts\": ${parameters.texts}, \"truncate\": ${parameters.truncate} }" , if user input "truncate": "123" in parameters when predict, will we pass \"truncate\": 123" to model with your PR?